File: Python/textbook_track/chapter07/ba7f.py
# Implement SmallParsimony
import itertools as it
import math
import sys

from collections import defaultdict

NUCLEOTIDES = ('A', 'C', 'G', 'T')


class Tree:
    def __init__(self, adjacency_list, nr_leaves):
        self._adjacency_list = adjacency_list
        self._nr_leaves = nr_leaves

        nr_nodes = max(self._adjacency_list.keys()) + 1
        self._nodes = tuple(range(nr_nodes))
        self._parents = self._calc_parents(nr_nodes)
        self._root = self._get_root_node_id()

    def _calc_parents(self, nr_nodes):
        parents = [-1] * nr_nodes
        for parent, children in self._adjacency_list.items():
            for child in children:
                parents[child] = parent
        return tuple(parents)

    def _get_root_node_id(self):
        for node, parent in enumerate(self._parents):
            if parent == -1:
                return node

    @property
    def nr_nodes(self):
        return len(self._nodes)

    @property
    def nodes(self):
        return self._nodes

    @property
    def root(self):
        return self._root

    def get_children(self, node):
        return self._adjacency_list.get(node, [])

    def get_parent(self, node):
        return self._parents[node]


def read_input_data(lines):
    leaf_id_generator = it.count(0)
    adjacency_list = defaultdict(list)
    characters = []
    for node_id, neighbour in map(lambda x: x.split("->"), lines):
        if neighbour.isnumeric():
            adjacency_list[int(node_id)].append(int(neighbour))
        else:
            adjacency_list[int(node_id)].append(next(leaf_id_generator))
            characters.append(neighbour)
    return dict(adjacency_list), characters


def calc_hamming_distance(s1, s2):
    distance = 0
    for c1, c2 in zip(s1, s2):
        if c1 != c2:
            distance += 1
    return distance


def initialize_minimum_parsimony_score_for_leaves(leaf_labels):
    minimum_parsimony_score = {}
    for leaf_id, label in enumerate(leaf_labels):
        minimum_parsimony_score[leaf_id] = tuple(0 if label == nucleotide else math.inf for nucleotide in NUCLEOTIDES)
    return minimum_parsimony_score


def get_ripe_nodes(tree, minimum_parsimony_scores):
    ripe_nodes = set()
    for node in tree.nodes:
        if children := tree.get_children(node):
            if all(child in minimum_parsimony_scores for child in children):
                ripe_nodes.add(node)
    return ripe_nodes


def calc_parsimony_scores_for_node(tree, minimum_parsimony_scores, node):
    (left_child, right_child) = tree.get_children(node)
    left_child_scores = minimum_parsimony_scores[left_child]
    right_child_scores = minimum_parsimony_scores[right_child]

    def calc_child_tree_score(node_label, child_scores):
        return min(score if char == node_label else score + 1 for char, score in zip(NUCLEOTIDES, child_scores))

    scores = []
    for nucleotide in NUCLEOTIDES:
        left = calc_child_tree_score(nucleotide, left_child_scores)
        right = calc_child_tree_score(nucleotide, right_child_scores)
        scores.append(left + right)

    return tuple(scores)


def get_most_parsimonious_labeling(tree, minimum_parsimony_scores):
    def get_label(scores):
        _, label = min(zip(scores, NUCLEOTIDES), key=lambda x: x[0])
        return label

    node_labels = [""] * tree.nr_nodes
    root_node = tree.root
    node_labels[root_node] = get_label(minimum_parsimony_scores[root_node])
    stack = list(tree.get_children(root_node))
    while stack:
        node_id = stack.pop(-1)
        parent_label = node_labels[tree.get_parent(node_id)]
        node_scores = minimum_parsimony_scores[node_id]
        modified_scores = [score if char == parent_label else score + 1
                           for char, score in zip(NUCLEOTIDES, node_scores)]
        node_labels[node_id] = get_label(modified_scores)
        stack.extend(tree.get_children(node_id))

    return node_labels


def solve_small_parsimony_for_single_label(tree, leaf_labels):
    minimum_parsimony_scores = initialize_minimum_parsimony_score_for_leaves(leaf_labels)
    ripe_nodes = get_ripe_nodes(tree, minimum_parsimony_scores)
    while ripe_nodes:
        node = ripe_nodes.pop()
        minimum_parsimony_scores[node] = calc_parsimony_scores_for_node(tree, minimum_parsimony_scores, node)
        if (parent_node := tree.get_parent(node)) != -1:
            children = tree.get_children(parent_node)
            if all(child in minimum_parsimony_scores for child in children):
                ripe_nodes.add(parent_node)

    return minimum_parsimony_scores


def solve_small_parsimony_problem(tree, characters):
    parsimony_score = 0
    node_labels = []
    for leaf_labels in zip(*characters):
        minimum_parsimony_scores = solve_small_parsimony_for_single_label(tree, leaf_labels)
        parsimony_score += min(minimum_parsimony_scores[tree.root])
        labels = get_most_parsimonious_labeling(tree, minimum_parsimony_scores)
        node_labels.append(labels)
    node_labels = tuple(map("".join, zip(*node_labels)))
    return parsimony_score, node_labels


def create_edge_output(tree, node_labels):
    for node_id in tree.nodes:
        node_label = node_labels[node_id]
        for child_id in tree.get_children(node_id):
            child_label = node_labels[child_id]
            distance = calc_hamming_distance(node_label, child_label)
            yield f"{node_label}->{child_label}:{distance}"
            yield f"{child_label}->{node_label}:{distance}"


def main():
    data = sys.stdin.read().splitlines()
    nr_leaves = int(data[0])
    adjacency_list, characters = read_input_data(data[1:])
    tree = Tree(adjacency_list, nr_leaves)
    parsimony_score, node_labels = solve_small_parsimony_problem(tree, characters)
    print(parsimony_score)
    edges = create_edge_output(tree, node_labels)
    for edge in edges:
        print(edge)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter07/BA7F.scala
package TextbookTrack.Chapter07

object BA7F {
  import scala.annotation.tailrec

  type Node = Int
  type Children = (Node, Node)

  private val Nucleotides: List[Char] = List('A', 'C', 'G', 'T')

  class Tree(adjacencyList: Map[Node, Children], val nrLeaves: Int) {
    val nrNodes: Int = adjacencyList.keysIterator.max + 1

    private val parents: Map[Node, Node] =
      adjacencyList.flatMap{ case (parent, (leftChild, rightChild)) => Map(leftChild -> parent, rightChild -> parent) }

    val root: Node = adjacencyList.keysIterator.find(!parents.contains(_)) match {
      case Some(node) => node
      case None => throw new Exception("Root of the tree is not found.")
    }

    def getParent(node: Node): Option[Node] = parents.get(node)

    def getChildren(node: Node): Option[Children] = adjacencyList.get(node)
  }

  private def readData(reader: Iterator[String]): (Map[Node, Children], List[String]) = {
    val leafIdGenerator: Iterator[Int] = Iterator.from(0)
    val lines: Iterator[(String, String)] = reader.map(_.split("->").toList).collect{ case List(a, b) => (a, b) }

    @tailrec
    def loop(edges: List[(Node, Node)], characters: List[String]): (Map[Node, Children], List[String]) =
      lines.nextOption() match {
        case None =>
          val adjacencyList: Map[Node, Children] =
            edges
              .groupMap{ case (nodeFrom, _) => nodeFrom}{ case (_, nodeTo) => nodeTo }
              .collect{ case (node, List(right, left)) => node -> (left, right) }
          (adjacencyList, characters.reverse)
        case Some((nodeId, b)) => b.toIntOption match {
          case None => loop((nodeId.toInt, leafIdGenerator.next()) :: edges, b :: characters)
          case Some(neighbourId) => loop((nodeId.toInt, neighbourId) :: edges, characters)
        }
    }

    loop(Nil, Nil)
  }

  private def calcHammingDistance(s1: String, s2: String): Int = s1.lazyZip(s2).count{ case (c1, c2) => c1 != c2 }

  private def initializeMinimumParsimonyScoreForLeaves(leafLabels: List[Char]): Map[Node, List[Double]] =
    leafLabels
      .zipWithIndex
      .map{
        case (label, leafId) =>
          leafId -> Nucleotides.map(nucleotide => if (nucleotide == label) 0.0 else Double.PositiveInfinity)
      }
      .toMap

  private def isRipe(tree: Tree, node: Node, minimumScores: Map[Node, List[Double]]): Boolean =
    tree.getChildren(node) match {
      case Some((left, right)) if minimumScores.contains(left) && minimumScores.contains(right) => true
      case _ => false
    }

  private def getRipeNodes(tree: Tree, leafScores: Map[Node, List[Double]]): List[Node] =
    (0 until tree.nrNodes).filter(isRipe(tree, _, leafScores)).toList

  private def calcParsimonyScoresForNode(tree: Tree,
                                         minimumParsimonyScores: Map[Node, List[Double]],
                                         node: Node): List[Double] = {
    def calcChildTreeScore(nodeLabel: Char, childScores: List[Double]): Double =
      Nucleotides.lazyZip(childScores).map{ case (char, score) => if (char == nodeLabel) score else score + 1 }.min

    tree.getChildren(node) match {
      case Some((left, right)) =>
        val leftChildScores: List[Double] = minimumParsimonyScores(left)
        val rightChildScores: List[Double] = minimumParsimonyScores(right)
        Nucleotides.map{ nucleotide =>
          calcChildTreeScore(nucleotide, leftChildScores) + calcChildTreeScore(nucleotide, rightChildScores)
        }
      case None => throw new Exception("Ripe node always have children.")
    }
  }

  private def solveSmallParsimonyForSingleLabel(tree: Tree, leafLabels: List[Char]): Map[Node, List[Double]] = {
    @tailrec
    def loop(minimumScores: Map[Node, List[Double]], ripeNodes: List[Node]): Map[Node, List[Double]] =
      ripeNodes match {
        case Nil => minimumScores
        case node :: rest =>
          val nodeScores: List[Double] = calcParsimonyScoresForNode(tree, minimumScores, node)
          val updatedMinimumScores: Map[Node, List[Double]] = minimumScores.updated(node, nodeScores)
          tree.getParent(node) match {
            case Some(parent) if isRipe(tree, parent, updatedMinimumScores) =>
              loop(updatedMinimumScores, parent :: rest)
            case _ => loop(updatedMinimumScores, rest)
          }
      }

    val leafScores: Map[Node, List[Double]] = initializeMinimumParsimonyScoreForLeaves(leafLabels)
    val initialRipeNodes: List[Node] = getRipeNodes(tree, leafScores)
    loop(leafScores, initialRipeNodes)
  }

  private def getMostParsimoniousLabeling(tree: Tree, minimumScores: Map[Int, List[Double]]): List[Char] = {
    def getLabel(scores: List[Double]): Char = scores.lazyZip(Nucleotides).minBy{ case (score, _) => score }._2

    val nodeLabels: Array[Char] = new Array[Char](tree.nrNodes)
    nodeLabels(tree.root) = getLabel(minimumScores(tree.root))

    @tailrec
    def loop(stack: List[Node]): List[Char] = stack match {
      case Nil => nodeLabels.toList
      case nodeId :: rest =>
        tree.getParent(nodeId) match {
          case None => loop(rest)
          case Some(parent) =>
            val parentLabel: Char = nodeLabels(parent)
            val nodeScores: List[Double] = minimumScores(nodeId)
            val modifiedScores: List[Double] =
              Nucleotides.lazyZip(nodeScores).map{ case (char, score) => if (char == parentLabel) score else score + 1 }
            nodeLabels(nodeId) = getLabel(modifiedScores)
            tree.getChildren(nodeId) match {
              case None => loop(rest)
              case Some((left, right)) => loop(left :: right :: rest)
            }
        }
    }

    val initialStack: List[Node] = tree.getChildren(tree.root) match {
      case None => Nil
      case Some((left, right)) => List(left, right)
    }
    loop(initialStack)
  }

  def solveSmallParsimonyProblem(tree: Tree, characters: List[String]): (Int, Vector[String]) = {
    val root: Node = tree.root
    val (totalScore, nodeLabels): (Double, List[List[Char]]) =
      characters.transpose.foldRight((0.0, Nil: List[List[Char]])){
        case (leafLabels, (score, nodeLabels)) =>
          val minimumScores: Map[Node, List[Double]] = solveSmallParsimonyForSingleLabel(tree, leafLabels)
          val labels: List[Char] = getMostParsimoniousLabeling(tree, minimumScores)
          (score + minimumScores(root).min, labels :: nodeLabels)
    }
    (totalScore.toInt, nodeLabels.transpose.map(_.mkString).toVector)
  }

  private def createEdgeOutput(tree: Tree, nodeLabels: Vector[String]): List[String] = {
    (0 until tree.nrNodes).flatMap{ nodeId =>
      val nodeLabel: String = nodeLabels(nodeId)
      tree.getChildren(nodeId).map{ case (left, right) => List(left, right) } match {
        case None => Nil
        case Some(children) =>
          children.flatMap{ child =>
            val childLabel: String = nodeLabels(child)
            val distance: Int = calcHammingDistance(nodeLabel, childLabel)
            List(s"$nodeLabel->$childLabel:$distance", s"$childLabel->$nodeLabel:$distance")
          }
      }
    }.toList
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrLeaves: Int = reader.next().toInt
    val (adjacencyList, characters): (Map[Node, Children], List[String]) = readData(reader)
    val tree = new Tree(adjacencyList, nrLeaves)
    val (parsimonyScore, nodeLabels): (Int, Vector[String]) = solveSmallParsimonyProblem(tree, characters)
    val edges: List[String] = createEdgeOutput(tree, nodeLabels)
    println(parsimonyScore)
    edges.foreach(println)
  }
}
